### PR TITLE
WT-9182 Only consider the size of updates being restored in sudp_memsize

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
This pr also contains the fix for WT-9592. I'll merge again after WT-9592 is in.

File: src/include/reconcile_inline.h
Comment (agorrod) [inline src/include/reconcile_inline.h:202]: It's worth noting that this calculation was pretty hard won. You can see the history in WT-6532 and WT-6487.

I suspect that the change here will mean that splits happen less often - which is the direction WT-6532 was going in anyway, so hopefully it's the right choice.

We don't have a direct test that will show a failure if this calculation changes, it will just alter the size of a database being created on disk, or the internal tree structure.

As such, this change deserves a set of sys-perf performance testing and analysis before merge.

File: src/include/reconcile.h
Comment (agorrod) [inline src/include/reconcile.h:229]: Would it be more accurate to call this the `supd_restore_memsize`?

File: src/reconcile/rec_write.c
Comment (agorrod) [inline src/reconcile/rec_write.c:1819]: I'm confused by this code. I think the code populating the list of saved updates, also populates the memory size. This code is trimming that list of saved updates down as the updates are distributed across different pages, but it's also re-incrementing the memory size for any updates that weren't removed.

Should it instead be decrementing the size of updates that have been associated with a different page? Or possibly not doing anything at all?

**Review by agorrod — COMMENTED** [review]
Thanks for talking me through this change Chenhao. I have a few questions and comments

File: src/include/reconcile_inline.h
Comment (quchenhao) [reply src/include/reconcile_inline.h:202]: Yes, I've run the sys-perf test. The result is in the ticket. The results are green and I'm not sure which test in the suite I should look at in detail though.

File: src/include/reconcile.h
Comment (quchenhao) [reply src/include/reconcile.h:229]: Yes, I'll change the name.

File: src/reconcile/rec_write.c
Comment (quchenhao) [reply src/reconcile/rec_write.c:1819]: I think what it does is to shift the saved updates that have been removed to the split page. In line 1805, it first reset the size to 0, so we should recalculate the size again.

File: src/include/reconcile.h
Comment (quchenhao) [reply src/include/reconcile.h]: [ce8c34b](https://github.com/wiredtiger/wiredtiger/pull/8127/commits/ce8c34bc0d6a24e0b3163090f32645b1290d157d)

File: src/reconcile/rec_write.c
Comment (agorrod) [reply src/reconcile/rec_write.c:1819]: I see - should it also be resetting `supd_restore_count` to 0?

Comment (quchenhao) [reply src/reconcile/rec_write.c:1820]: Good point. Fixed in [d4aecbe](https://github.com/wiredtiger/wiredtiger/pull/8127/commits/d4aecbeddf100340aaaa98f022cc040aed90083f)

**Comment by agorrod** [comment]
I'm somewhat consciously sitting on this Chenhao - with the segfault and other issues floating around, I'm reluctant to introduce this right now.
Am I misreading the urgency?  I know it's annoying to leave something pending, but I'm also a bit worried about fallout from this one.

**Comment by quchenhao** [comment]
@agorrod It's not urgent. I agree we should leave this for now until evergreen is stable.

File: src/include/reconcile_inline.h
Comment (quchenhao) [inline src/include/reconcile_inline.h:202]: Should we skip this optimization for in memory database? I feel there is no benefit to split for in memory case as it only leads us to use more memory.

**Comment by quchenhao** [comment]
@agorrod @kommiharibabu This is ready for review. I still have a question about in memory database. Should we skip the optimization to encourage split for in memory database? I feel it doesn't provide much benefit in the in memory case and it only leads to more memory consumption as we have a deeper tree and more page overhead.

File: src/reconcile/rec_write.c
Comment (quchenhao) [reply src/reconcile/rec_write.c:1831]: No r->supd_restore_count is at the update chain level not at the updates level. It is updated in line 1814

**Comment by quchenhao** [comment]
@agorrod Please have a look at this again when you have time.

File: src/reconcile/rec_write.c
Comment (kommiharibabu) [reply src/reconcile/rec_write.c:1831]: Thanks.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by quchenhao** [comment]
@agorrod I have added the config to enable this optimization. Can you have a look?

**Comment by agorrod** [comment]
@quchenhao, now that we have moved this work to the backlog, would you mind recording the current state in the JIRA ticket and closing this PR?

**Comment by quchenhao** [comment]
Close to work in the future

**Pull request closed without merging**
